docs/guides/docker_image_with_your_release.md

```diff
@@ -0,0 +1,74 @@
+# Build a docker image with your release
+
+Now that we have seen how to build a release, let's build one in a docker image. This is necessary for a deployment with docker swarm or kubernetes for example.
+
+A prerequisite step to this, is to initialize your release (`mix release.init`) and configure your [vm.args](/files/vm.args).
+
+in order to customize the application release version when building the docker image, change the following in your `mix.exs` file
+
+```
+# add this
+@version System.get_env("APP_VERSION") || "0.0.0"
```

Comment:
In general I find it easier to use the `:version` field in `mix.exs` normally, and grep it out of `mix.exs` in a Makefile which is used to build images, i.e.:

```mix.exs
def project do
  # ...
  version: "0.1.0"
```

```make
IMAGE ?= `grep 'app:' mix.exs | sed -e 's/ //g' -e 's/app://' -e 's/[:,]//g'`
VERSION ?= `grep 'version' mix.exs | sed -e 's/ //g' -e 's/version://' -e 's/[",]//g'`

#...
image: ## Build Docker image
    docker build -t $(IMAGE):$(VERSION) .
```

This is portable to any project, and will build a tagged, versioned image of the form `myapp:0.1.0` by running `make image` in the project root.

Reply:
thanks for the tip with grep and sed.
What I liked about this, is that it doesn't force you to commit a new version number everytime you make changes to the code. I feel almost that the version number is a config and shouldn't appear in the code. I understand it's a matter of taste.

I'll replace this with this new way then.

Reply:
My take on that is that you should append a git shorthash to the version number if you are building new images while not changing the version - this keeps the application version the same, but the release artifact gets a distinct "release" version:

```make
BUILD ?= `git describe --always`
IMAGE ?= `grep 'app:' mix.exs | sed -e 's/ //g' -e 's/app://' -e 's/[:,]//g'`
VERSION ?= `grep 'version' mix.exs | sed -e 's/ //g' -e 's/version://' -e 's/[",]//g'`

image: ## Build Docker image
    docker build -t $(IMAGE):$(VERSION)-$(BUILD)
```

Reply:
One question regarding the BUILD variable, shouldn't it be `git rev-parse --short HEAD` ?

`git describe --always` gives me a full version number like `v0.0.1-355-g97f75e69`
(I have to admit, I'm not super familiar with makefiles, so maybe there is something I am missing)

my second question relates to the build stage of a continuous integration agent. It would look something like this then ?

```
BUILD= `git rev-parse --short HEAD`
            IMAGE= `grep 'app:' mix.exs | sed -e 's/ //g' -e 's/app://' -e 's/[:,]//g'`
            VERSION= `grep 'version' mix.exs | sed -e 's/ //g' -e 's/version://' -e 's/[",]//g'`
            make
            docker login -u $DOCKER_LOGIN -p $DOCKER_PASSWORD
            docker push dockerhub_username/$IMAGE:$VERSION-$BUILD
            docker push dockerhub_username/$IMAGE:latest
```

Reply:
for anybody else interested, make wasn't directly available on circle_ci machine so I ended going with the following
```
echo "export BUILD=$(git rev-parse --short HEAD)" >> $BASH_ENV
echo "export IMAGE=$(grep 'app:' mix.exs | sed -e 's/ //g' -e 's/app://' -e 's/[:,]//g')" >> $BASH_ENV
echo "export VERSION=$(grep 'version' mix.exs | sed -e 's/ //g' -e 's/version://' -e 's/[",]//g')" >> $BASH_ENV
source $BASH_ENV
docker build -t happysada/$IMAGE:$VERSION-$BUILD -t happysalada/$IMAGE:latest .
docker login -u $DOCKER_LOGIN -p $DOCKER_PASSWORD
docker push happysalada/$IMAGE:$VERSION-$BUILD
docker push happysalada/$IMAGE:latest
```